elements/lisk-p2p/src/peer/outbound.ts

```diff
@@ -0,0 +1,340 @@
+} from '../p2p_types';
+
+import * as querystring from 'querystring';
+import * as socketClusterClient from 'socketcluster-client';
+import { validatePeerInfo } from '../validation';
+
+type SCClientSocket = socketClusterClient.SCClientSocket;
+
+export const EVENT_CONNECT_OUTBOUND = 'connectOutbound';
+export const EVENT_CONNECT_ABORT_OUTBOUND = 'connectAbortOutbound';
+export const EVENT_CLOSE_OUTBOUND = 'closeOutbound';
+export const EVENT_OUTBOUND_SOCKET_ERROR = 'outboundSocketError';
+
+export interface PeerInfoAndOutboundConnection {
+	readonly peerInfo: P2PDiscoveredPeerInfo;
+	readonly socket: SCClientSocket;
+}
+
+export class OutboundPeer extends Peer {
+	protected _socket: SCClientSocket | undefined;
+
+	public constructor(
+		peerInfo: P2PDiscoveredPeerInfo,
+		peerSocket?: SCClientSocket,
+		peerConfig?: PeerConfig,
+	) {
+		super(peerInfo, peerConfig);
+		this._socket = peerSocket ? peerSocket : undefined;
+		if (this._socket) {
+			this._bindHandlersToOutboundSocket(this._socket);
+		}
+	}
+
+	public set socket(scClientSocket: SCClientSocket) {
+		if (this._socket) {
+			this._unbindHandlersFromOutboundSocket(this._socket);
+		}
+		this._socket = scClientSocket;
+		this._bindHandlersToOutboundSocket(this._socket);
+	}
+
+	public send(packet: P2PMessagePacket): void {
+		if (!this._socket) {
+			this._socket = this._createOutboundSocket();
+		}
+
+		super.send(packet);
+	}
+
+	public async request(packet: P2PRequestPacket): Promise<P2PResponsePacket> {
+		if (!this._socket) {
+			this._socket = this._createOutboundSocket();
+		}
+
+		return super.request(packet);
+	}
+
+	private _createOutboundSocket(): SCClientSocket {
+		const legacyNodeInfo = this._nodeInfo
+			? convertNodeInfoToLegacyFormat(this._nodeInfo)
+			: undefined;
+
+		const connectTimeout = this._peerConfig.connectTimeout
+			? this._peerConfig.connectTimeout
+			: DEFAULT_CONNECT_TIMEOUT;
+		const ackTimeout = this._peerConfig.ackTimeout
+			? this._peerConfig.ackTimeout
+			: DEFAULT_ACK_TIMEOUT;
+
+		// Ideally, we should JSON-serialize the whole NodeInfo object but this cannot be done for compatibility reasons, so instead we put it inside an options property.
+		const clientOptions: ClientOptionsUpdated = {
+			hostname: this._ipAddress,
+			port: this._wsPort,
+			query: querystring.stringify({
+				...legacyNodeInfo,
+				options: JSON.stringify(legacyNodeInfo),
+			}),
+			connectTimeout,
+			ackTimeout,
+			multiplex: false,
+			autoConnect: false,
+			autoReconnect: false,
+			pingTimeoutDisabled: true,
+		};
+
+		const outboundSocket = socketClusterClient.create(clientOptions);
+
+		this._bindHandlersToOutboundSocket(outboundSocket);
+
+		return outboundSocket;
+	}
+
+	public connect(): void {
+		if (!this._socket) {
+			this._socket = this._createOutboundSocket();
+		}
+		this._socket.connect();
+	}
+
+	public disconnect(code: number = 1000, reason?: string): void {
+		if (!this._socket) {
+			throw new Error('Peer socket does not exist');
```

Comment:
Previously, when disconnecting the peer, if the outbound socket did not exist, this was a no-op (fail-safe).
If we throw an error here, then the peer_pool will need to check if the socket exists on the peer before invoking disconnect (which we are not doing now).

I think it's better in this case to make it a no-op if the socket does not exist (since the absence of the socket means that it's disconnected already so it's not exactly a failure).